[17.0][IMP] product_contract: Add contract configurator instead of making tree not editable

carolinafernandez-tecnativa opened the pull request:
Before these changes, when trying to edit a line of sale order, it was opening the form of the line. But following the Odoo way of working with the events sales, we have make a new contract configurator that will be opened in a popup when selecting a contract product.

@Tecnativa TT50535

pilarvargas-tecnativa reviewed (approved):
:+1:

carolinafernandez-tecnativa commented:
ping @pedrobaeza

pedrobaeza commented on product_contract/views/sale_order.xml:
```diff
@@ -95,17 +95,64 @@
+                <field
+                    name="contract_id"
+                    options='{"no_create": True}'
+                    column_invisible="parent.is_contract == False"
```
Shorter, but also more correct, as `is_contract` may be other value than `False`, but also a falsy value.

```suggestion
                    column_invisible="not parent.is_contract"
```

Do the same on the rest.

pedrobaeza commented on product_contract/views/sale_order.xml:
```diff
@@ -95,17 +95,64 @@
                 <field
                     name="date_start"
                     column_invisible="parent.is_contract == False"
+                    required="is_contract == True"
```
```suggestion
                    required="is_contract"
```

and the same for the rest of the occurrences.

pedrobaeza reviewed (changes requested):
I'm not finding the corresponding fields in the sales order line tree:

![image](https://github.com/user-attachments/assets/2c1c561c-f536-4102-ae19-89db610d92a9)


Anyway, seeing how Odoo has adapted other similar needs, where we need extra info for a sales order line, showing a popup with that extra details, I think it should be the way to go:

![Peek 2024-08-16 13-30](https://github.com/user-attachments/assets/3ed387ce-58bb-400d-868a-725e521841b8)

It seems not very complicated to be added, taking `event_sale` as example:

https://github.com/odoo/odoo/tree/17.0/addons/event_sale/static/src/js

@CarlosRoca13 can you please take this and perform it, as you already knows JS and the event_sale controller?

@sbejaoui do you agree?

Another question: why not doing a `detailed_type='contract'` instead of having a special field `is_contract` at product level?

CarlosRoca13 commented:
@pedrobaeza @carolinafernandez-tecnativa pushed changes to add the product contract configurator like odoo does for events

carolinafernandez-tecnativa commented:
@pedrobaeza LGTM, what do you think?

pedrobaeza commented:
The PR title is incorrect, and I think the first commit is now not needed if it has to be undo in the second one.

pedrobaeza reviewed (commented):
About the functionality, this sounds better, although the "Contract" and "Contract line recurrency" should be at the end in a separate section, as this a very special case. I'm also surprised that the "End date" is filled by default, where it shouldn't.

And what about adding some demo master data to ease the tests?

CarlosRoca13 committed:
[IMP] product_contract: Add contract configurator instead of making tree not editable

Before this changes, when trying to edit a line of sale order, it was
opening the form of the line. But following the way to work of odoo
with sale event, we have make a new contract configurator that will
be opened when selecting a product of type contract.

CarlosRoca13 committed:
[IMP] product_contract: Add posibility to compute date_start of line using confirmation date_start

With these changes, we allow the contract line start date to be computed
using the order confirmation date. When the product is configured with
any of the options set in contract_start_date_method other than manual,
the start date will be calculated based on the established date and the
selected period.

Additionally, we can force the month in which we will work in case the
frequency is yearly, quarterly, or semesterly.

Is not added support for daily, weekly or monthlylastday in this commit.

pedrobaeza commented on product_contract/models/product_template.py:
```diff
@@ -61,6 +61,71 @@ class ProductTemplate(models.Model):
+        [
+            ("1", "1"),
+            ("2", "2"),
+            ("3", "3"),
```
```suggestion
            ("1", "First month"),
            ("2", "Second month"),
            ("3", "Third month"),
```

pedrobaeza commented on product_contract/models/product_template.py:
```diff
@@ -61,6 +61,71 @@ class ProductTemplate(models.Model):
+            ("2", "2"),
+            ("3", "3"),
+        ],
+        "Force Month",
```
```suggestion
        "Force Month",
        help="Force the month to be used inside the quarter",
```

pedrobaeza commented on product_contract/models/product_template.py:
```diff
@@ -61,6 +61,71 @@ class ProductTemplate(models.Model):
+            ("3", "3"),
+            ("4", "4"),
+            ("5", "5"),
+            ("6", "6"),
```
The same here about the labels and the help.

pedrobaeza commented on product_contract/models/sale_order_line.py:
```diff
@@ -269,3 +283,53 @@ def _compute_qty_to_invoice(self):
+                if is_end:
+                    start_date = (
+                        start_date
+                        + relativedelta(day=1, months=1)
+                        - relativedelta(days=1)
```
Does this work?

```suggestion
                        + relativedelta(day=1, months=1, days=-1)
```

pedrobaeza commented on product_contract/models/sale_order_line.py:
```diff
@@ -269,3 +283,53 @@ def _compute_qty_to_invoice(self):
+                increment = month_nb if not forced_month else 0
+                start_date = (
+                    start_date
+                    + relativedelta(months=1 + increment)
```
Isn't this missing?

```suggestion
                    + relativedelta(months=1 + increment, day=1)
```

CarlosRoca13 replied:
Already established in the above definition https://github.com/OCA/contract/pull/1120/files#diff-308d8794eb63f4f473a8ce812eba0aacdbdf86a8c0c8f9c3786f1de1baade116R317-R319

pedrobaeza commented on product_contract/readme/ROADMAP.md:
```diff
@@ -0,0 +1,2 @@
+Is not added support for dates computed from confirmation date of sale order for
+recurrent types daily, weekly or monthlylastday
```
```suggestion
- There's no support right now for computing the start date for the
  following recurrent types: daily, weekly and monthlylastday.
```

pedrobaeza commented on product_contract/models/sale_order_line.py:
```diff
@@ -269,3 +283,53 @@ def _compute_qty_to_invoice(self):
+            today = fields.Date.today()
+            month_period = month = today.month
+            month_nb = MONTH_NB_MAPPING[line.recurring_rule_type]
+            # The period number is started by 0 to be able calculate the month
```
```suggestion
            # The period number is started by 0 to be able to calculate the month
```

pedrobaeza commented on product_contract/wizards/product_contract_configurator.py:
```diff
@@ -0,0 +1,127 @@
+        string="Contract Template",
+        compute="_compute_contract_template_id",
+    )
+    recurring_rule_type = fields.Selection(
```
Now that we have configured all the start date thing in the product, it doesn't seem OK to also have again this configuration here, don't you think?

CarlosRoca13 replied:
I think we should keep it in case the product is set to manual mode, but the unnecessary fields should be hidden in the view.

pedrobaeza replied:
@sbejaoui which was the reason to define a recurrency in the product and ask for that recurrency again in the sales order line? Can we strip that part for simplifying code and UI?

CarlosRoca13 commented:
Changes done. Let me add some tests for this new feature

CarlosRoca13 commented:
Tests added

pedrobaeza reviewed (commented):
Seems to work OK, but I find something weird, as if you add any line in the contract template, when confirming the quotation, both the sales order line and the template lines are added in the resulting contract. Was it that way in the original 12.0 implementation?

Something that may be related to this PR is that the extras lines don't have the proper start date.

pedrobaeza commented:
Ah, another thing: I think that the extra description about the recurrency and so on shouldn't be transferred to the contract line. Maybe you can add some markers at the beginning and the end to easily recognize it and remove it when creating the contract line.

CarlosRoca13 commented:
> Was it that way in the original 12.0 implementation?

Yes, when the template has a line that line is added to the contract.

> Something that may be related to this PR is that the extras lines don't have the proper start date.

What date should be set to that lines?

> I think that the extra description about the recurrency and so on shouldn't be transferred to the contract line. Maybe you can add some markers at the beginning and the end to easily recognize it and remove it when creating the contract line.

Ok, I'll let you know when it's ready.

CarlosRoca13 commented:
Change done

pedrobaeza commented:
> What date should be set to that lines?

The same that was computed for the sales order line.

pedrobaeza reviewed (approved):
/ocabot merge major
